Mathlib/Data/Real/Pi/Irrational.lean

```diff
@@ -0,0 +1,248 @@
+      (cosPoly n).eval₂ (Int.castRingHom _) θ * cos θ)
+  | 0 => by simp [sinPoly, cosPoly, I_zero, eval₂_C]
+  | 1 => by
+      simp only [I_one, cosPoly, sinPoly, Nat.factorial_one, Nat.cast_one, mul_one, eval₂_C,
+        eval₂_monomial, Nat.reduceAdd]
+      simp [sub_eq_add_neg]
+  | (n+2) => by
+      rw [Nat.mul_succ, add_right_comm, add_comm (_ + _), pow_add, ← mul_assoc, recursion, sub_mul,
+        mul_assoc, mul_assoc _ (I n θ), sinPoly_add_cosPoly_eval, mul_add_one 2, add_right_comm,
+        pow_add, ← mul_assoc (I n θ), sinPoly_add_cosPoly_eval, sinPoly, cosPoly, eval₂_add,
+        eval₂_add, eval₂_smul, eval₂_smul, eval₂_mul, eval₂_mul, eval₂_monomial]
+      simp only [map_mul, map_ofNat, map_add, map_natCast, Int.reduceNeg, map_neg, neg_mul,
+        Nat.factorial_succ]
+      push_cast
+      ring
+
+open BigOperators
+
+private lemma is_integer {p : Polynomial ℤ} (a b : ℤ) {k : ℕ} (hp : p.natDegree ≤ k) :
+    ∃ z : ℤ, p.eval₂ (Int.castRingHom ℝ) (a / b) * b ^ k = z := by
+  rcases eq_or_ne b 0 with rfl | hb
+  · rcases k.eq_zero_or_pos with rfl | hk
+    · exact ⟨p.coeff 0, by simp⟩
+    exact ⟨0, by simp [hk.ne']⟩
+  refine ⟨∑ i in p.support, p.coeff i * a ^ i * b ^ (k - i), ?_⟩
+  conv => lhs; rw [← sum_monomial_eq p]
+  rw [eval₂_sum, Polynomial.sum, Finset.sum_mul, Int.cast_sum]
+  simp only [eval₂_monomial, eq_intCast, div_pow, Int.cast_mul, Int.cast_pow]
+  refine Finset.sum_congr rfl (fun i hi => ?_)
+  have ik := (le_natDegree_of_mem_supp i hi).trans hp
+  rw [mul_assoc, div_mul_comm, ← Int.cast_pow, ← Int.cast_pow, ← Int.cast_pow, ← pow_sub_mul_pow b ik,
+    ← Int.cast_div_charZero, Int.mul_ediv_cancel _ (pow_ne_zero _ hb), ← mul_assoc, mul_right_comm,
+    ← Int.cast_pow]
+  exact dvd_mul_left _ _
+
+open Filter
+
+private lemma I_pos : 0 < I n (π / 2) := by
+  refine integral_pos (by norm_num) (Continuous.continuousOn (by continuity)) ?_ ⟨0, by simp⟩
+  refine fun x hx => mul_nonneg (pow_nonneg ?_ _) ?_
+  · rw [sub_nonneg, sq_le_one_iff_abs_le_one, abs_le]
+    exact ⟨hx.1.le, hx.2⟩
+  refine cos_nonneg_of_neg_pi_div_two_le_of_le ?_ ?_ <;>
+  nlinarith [hx.1, hx.2, pi_pos]
+
+private lemma I_le (n : ℕ) : I n (π / 2) ≤ 2 := by
+  rw [← norm_of_nonneg I_pos.le]
+  refine (norm_integral_le_of_norm_le_const ?_).trans (show (1 : ℝ) * _ ≤ _ by norm_num)
+  intros x hx
+  simp only [uIoc_of_le, neg_le_self_iff, zero_le_one, mem_Ioc] at hx
+  rw [norm_eq_abs, abs_mul, abs_pow]
+  refine mul_le_one (pow_le_one _ (abs_nonneg _) ?_) (abs_nonneg _) (abs_cos_le_one _)
+  rw [abs_le]
+  constructor <;>
+  nlinarith
+
+private lemma my_tendsto_pow_div_factorial_at_top (a : ℝ) :
+  Tendsto (fun n => (a : ℝ) ^ (2 * n + 1) / n !) atTop (nhds 0) := by
+  rw [← mul_zero a]
+  refine ((tendsto_pow_div_factorial_atTop (a ^ 2)).const_mul a).congr (fun x => ?_)
+  rw [← pow_mul, mul_div_assoc', _root_.pow_succ']
+
+private lemma not_irrational.exists_rep {x : ℝ} :
```

Comment:
This feels useful enough to be somewhere else.

Reply:
I think I'd make a slightly different claim, that this is useful enough that we probably already have the result somewhere else.

Reply:
I just looked again and I can't find anything quite right - `irrational_iff_ne_rational` is probably the one you expected to exist, but that's not close enough here because I want a positive natural denominator.